javaslang-benchmark/src/test/java/javaslang/idiom/ForBenchmark.java

```diff
@@ -0,0 +1,122 @@
+        JmhRunner.runDebugWithAsserts(CLASSES);
+    }
+
+    public static void main(String... args) {
+        JmhRunner.runNormalNoAsserts(CLASSES);
+    }
+
+    @State(Scope.Benchmark)
+    public static class Base {
+        @Param({ "10", "100", "1000" })
+        public int CONTAINER_SIZE;
+
+        Integer[] ELEMENTS;
+        List<Integer> ELEMENTS_ARRAY;
+
+        @Setup
+        public void setup() {
+            ELEMENTS = getRandomValues(CONTAINER_SIZE, 0);
+            ELEMENTS_ARRAY = Arrays.asList(ELEMENTS);
+        }
+    }
+
+    public static class ForComprehension1Iterator extends Base {
+
+        @Benchmark
+        public Object javaslang_for() {
+            final Iterable<Integer> result = For(ELEMENTS_ARRAY)
+                    .yield(i -> i * 0);
+
+            assert Iterator.ofAll(result).forAll(e -> e == 0);
+            return result;
+        }
+
+        @Benchmark
+        public Object java_for() {
+            final List<Integer> result = new ArrayList<>(ELEMENTS_ARRAY.size());
+            for (Integer i : ELEMENTS_ARRAY) {
+                result.add(i * 0);
+            }
+
+            assert Iterator.ofAll(result).forAll(e -> e == 0);
+            return result;
+        }
+    }
+
+    public static class ForComprehension2Iterators extends Base {
+
+        @Benchmark
+        public Object javaslang_for() {
+            final Iterable<Integer> result = For(ELEMENTS_ARRAY, ELEMENTS_ARRAY, ELEMENTS_ARRAY)
+                    .yield((i, j, k) -> i * j * k * 0);
+
+            assert Iterator.ofAll(result).forAll(e -> e == 0);
+            return result;
+        }
+
+        @Benchmark
+        public Object java_for() {
+            final List<Integer> result = new ArrayList<>(ELEMENTS_ARRAY.size() * ELEMENTS_ARRAY.size());
+            for (Integer i : ELEMENTS_ARRAY) {
+                for (Integer j : ELEMENTS_ARRAY) {
+                    result.add(i * j * 0);
+                }
+            }
+
+            assert Iterator.ofAll(result).forAll(e -> e == 0);
+            return result;
+        }
+    }
+
+    public static class ForComprehension3Iterators extends Base {
+
+        @Benchmark
+        public Object javaslang_for() {
+            final Iterable<Integer> result = For(ELEMENTS_ARRAY, ELEMENTS_ARRAY)
+                    .yield((i, j) -> i * j * 0);
+
+            assert Iterator.ofAll(result).forAll(e -> e == 0);
+            return result;
+        }
+
+        @Benchmark
+        public Object java_for() {
+            final List<Integer> result = new ArrayList<>(ELEMENTS_ARRAY.size() * ELEMENTS_ARRAY.size() * ELEMENTS_ARRAY.size());
+            for (Integer i : ELEMENTS_ARRAY) {
+                for (Integer j : ELEMENTS_ARRAY) {
+                    for (Integer k : ELEMENTS_ARRAY) {
+                        result.add(i * j * k * 0);
+                    }
+                }
+            }
+
+            assert Iterator.ofAll(result).forAll(e -> e == 0);
```

Comment:
I don't really like where all this is going, neither the fact that the build isn't even passing (again, please run it locally before committing).
These tests should be a *LOT* simpler and a lot more intuitive, otherwise we don't know what to say about the results.